crates/engine/local/src/payload.rs

```diff
@@ -2,29 +2,62 @@
 //! [`LocalEngineService`](super::service::LocalEngineService).
 
 use alloy_primitives::{Address, B256};
+use reth_chainspec::EthereumHardforks;
 use reth_ethereum_engine_primitives::EthPayloadAttributes;
 use reth_payload_primitives::PayloadAttributesBuilder;
-use std::{convert::Infallible, time::UNIX_EPOCH};
+use std::sync::Arc;
 
 /// The attributes builder for local Ethereum payload.
 #[derive(Debug)]
-pub struct EthLocalPayloadAttributesBuilder;
-
-impl PayloadAttributesBuilder for EthLocalPayloadAttributesBuilder {
-    type PayloadAttributes = EthPayloadAttributes;
-    type Error = Infallible;
+#[non_exhaustive]
+pub struct LocalPayloadAttributesBuilder<ChainSpec> {
+    chain_spec: Arc<ChainSpec>,
+}
 
-    fn build(&self) -> Result<Self::PayloadAttributes, Self::Error> {
-        let ts = std::time::SystemTime::now()
-            .duration_since(UNIX_EPOCH)
-            .expect("cannot be earlier than UNIX_EPOCH");
+impl<ChainSpec> LocalPayloadAttributesBuilder<ChainSpec> {
+    /// Creates a new instance of the builder.
+    pub const fn new(chain_spec: Arc<ChainSpec>) -> Self {
+        Self { chain_spec }
+    }
+}
 
-        Ok(EthPayloadAttributes {
-            timestamp: ts.as_secs(),
+impl<ChainSpec> PayloadAttributesBuilder<EthPayloadAttributes>
+    for LocalPayloadAttributesBuilder<ChainSpec>
+where
+    ChainSpec: Send + Sync + EthereumHardforks + 'static,
+{
+    fn build(&self, timestamp: u64) -> EthPayloadAttributes {
+        EthPayloadAttributes {
+            timestamp,
             prev_randao: B256::random(),
             suggested_fee_recipient: Address::random(),
-            withdrawals: None,
-            parent_beacon_block_root: None,
-        })
+            withdrawals: if self.chain_spec.is_shanghai_active_at_timestamp(timestamp) {
+                Some(Default::default())
+            } else {
+                None
+            },
+            parent_beacon_block_root: if self.chain_spec.is_cancun_active_at_timestamp(timestamp) {
+                Some(B256::random())
+            } else {
+                None
+            },
+        }
+    }
+}
+
+#[cfg(feature = "optimism")]
+impl<ChainSpec> PayloadAttributesBuilder<op_alloy_rpc_types_engine::OpPayloadAttributes>
+    for LocalPayloadAttributesBuilder<ChainSpec>
+where
+    ChainSpec: Send + Sync + EthereumHardforks + 'static,
+{
+    fn build(&self, timestamp: u64) -> op_alloy_rpc_types_engine::OpPayloadAttributes {
+        op_alloy_rpc_types_engine::OpPayloadAttributes {
+            payload_attributes: self.build(timestamp),
+            transactions: None,
+            no_tx_pool: None,
```

Comment:
this should work because the default will be false, so this would mean the node will be in sequencer mode.